### PR TITLE
Added typescript support, and refactored the dist folder

### PR DESCRIPTION
**Pull request by frankm1870**

### CODE REVIEW

File: build.js [hunk 1]
```diff
@@ -120,32 +122,35 @@ await remove("./dist", {
 	force: true,
 });
 
+await copy("./src", "./dist", {
```
Comment (Naviary2) [inline build.js:125]: Please bring the comment from below to here, and modify it according to the new context.

File: build.js [hunk 2]
```diff
@@ -120,32 +122,35 @@ await remove("./dist", {
...
+		console.error('TypeScript compilation failed with the following error:');
+		console.log(e.stdout.toString()); // Print TypeScript error output
+		console.log(e.stderr.toString()); // Print additional error details if available
+	}
+}
 
 if (!BUNDLE_FILES) {
 	/**
 	 * In development, copy all clientside files directly over to /dist AS THEY ARE.
 	 * No files are bundled. This means a LOT of requests! But, all our comments are there!
```
Comment (Naviary2) [inline build.js:144]: This is the comment that needs to be moved above and modified to match the context.

Also this block is completely empty and no longer needed.

**Comment by Naviary2** [comment]
I can't think of anything major I'd change, this is really solid!

I still want to do a little more bug searching before I merge this, and please do make the fix suggested above.

Also I did make a couple (very) minor changes so you may want to refresh your branch before making more changes.

**Comment by Naviary2** [comment]
Also this weird file `tsconfig.tsbuildinfo` keeps generating, wth is that supposed to be

**Comment by Naviary2** [comment]
My Intellisense works cross-file too so 🔥
We're gonna be able to write typescript alongside javascript shortly!

File: build.js [hunk 3]
```diff
@@ -120,32 +122,35 @@ await remove("./dist", {
 	force: true,
 });
 
+await copy("./src", "./dist", {
+	recursive: true,
+	force: true
+});
+
+if ((await getAllFilesInDirectoryWithExtension("./dist", ".ts")).length !== 0) {
```
[the remark this answers is not in the document]
Comment (frankm1870) [reply build.js:130]: The compilation fails; I don't know why, but it doesn't like trying to compile when there are no files to compile.

[the remark this answers is not in the document]
Comment (Naviary2) [reply build.js]: So it's not a github issue, it happens in your Visual Studio?

**Comment by Naviary2** [comment]
Thanks for your work! Ready to be merged

**Pull request merged**